File: backend/app/routers/agents.py
"""
Agents router — custom AI agents marketplace and configuration.
"""

from __future__ import annotations

import logging
import uuid
from datetime import datetime, timezone
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy import select, or_, and_
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user
from app.models.users import User
from app.models.agents import Agent
from app.schemas.agents import AgentCreate, AgentUpdate, AgentResponse

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/agents", tags=["agents"])


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


@router.get("", response_model=List[AgentResponse])
async def list_agents(
    category: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> List[AgentResponse]:
    """List public and author-owned AI agents."""
    filters = [
        or_(
            Agent.is_public == True,  # noqa: E712
            Agent.author_id == current_user.id,
        )
    ]

    if category and category != "all":
        filters.append(Agent.category == category)

    result = await db.execute(
        select(Agent)
        .where(and_(*filters))
        .order_by(Agent.usage_count.desc(), Agent.created_at.desc())
    )
    agents = result.scalars().all()

    response_list = []
    for a in agents:
        response_list.append(
            AgentResponse(
                id=a.id,
                name=a.name,
                description=a.description,
                modelId=a.model_id,
                systemPrompt=a.system_prompt,
                category=a.category,
                isPublic=a.is_public,
                rating=a.rating,
                usageCount=a.usage_count,
                createdBy=a.author_id,
                iconEmoji=a.icon_url or "🤖",
                tags=["AI", a.category] if a.category else ["AI"],
                createdAt=a.created_at,
            )
        )
    return response_list


@router.post("", response_model=AgentResponse, status_code=status.HTTP_201_CREATED)
async def create_agent(
    payload: AgentCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> AgentResponse:
    """Create a new custom prompt-engineered AI agent."""
    agent_id = uuid.uuid4()

    agent = Agent(
        id=agent_id,
        name=payload.name,
        description=payload.description,
        author_id=current_user.id,
        model_id=payload.model_id,
        system_prompt=payload.system_prompt,
        category=payload.category,
        is_public=payload.is_public,
        is_featured=False,
        usage_count=0,
        rating=5.0,
        price_credits=0,
        icon_url=payload.icon_emoji or "🤖",
        created_at=_utcnow(),
        updated_at=_utcnow(),
    )
    db.add(agent)
    await db.commit()
    await db.refresh(agent)

    return AgentResponse(
        id=agent.id,
        name=agent.name,
        description=agent.description,
        modelId=agent.model_id,
        systemPrompt=agent.system_prompt,
        category=agent.category,
        isPublic=agent.is_public,
        rating=agent.rating,
        usageCount=agent.usage_count,
        createdBy=agent.author_id,
        iconEmoji=agent.icon_url or "🤖",
        tags=["AI", agent.category],
        createdAt=agent.created_at,
    )


@router.patch("/{agent_id}", response_model=AgentResponse)
async def update_agent(
    agent_id: uuid.UUID,
    payload: AgentUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> AgentResponse:
    """Update configurations of a custom AI agent."""
    agent = await db.get(Agent, agent_id)
    if not agent:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="AI Agent not found.",
        )

    if agent.author_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only the author can update this AI agent.",
        )

    if payload.name is not None:
        agent.name = payload.name.strip()
    if payload.description is not None:
        agent.description = payload.description.strip()
    if payload.model_id is not None:
        agent.model_id = payload.model_id
    if payload.system_prompt is not None:
        agent.system_prompt = payload.system_prompt
    if payload.category is not None:
        agent.category = payload.category
    if payload.is_public is not None:
        agent.is_public = payload.is_public
    if payload.icon_emoji is not None:
        agent.icon_url = payload.icon_emoji

    agent.updated_at = _utcnow()
    db.add(agent)
    await db.commit()
    await db.refresh(agent)

    return AgentResponse(
        id=agent.id,
        name=agent.name,
        description=agent.description,
        modelId=agent.model_id,
        systemPrompt=agent.system_prompt,
        category=agent.category,
        isPublic=agent.is_public,
        rating=agent.rating,
        usageCount=agent.usage_count,
        createdBy=agent.author_id,
        iconEmoji=agent.icon_url or "🤖",
        tags=["AI", agent.category],
        createdAt=agent.created_at,
    )


@router.delete("/{agent_id}", status_code=status.HTTP_204_NO_CONTENT, response_class=Response)
async def delete_agent(
    agent_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Response:
    """Delete a custom AI agent."""
    agent = await db.get(Agent, agent_id)
    if not agent:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="AI Agent not found.",
        )

    if agent.author_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only the author can delete this AI agent.",
        )

    await db.delete(agent)
    await db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: backend/src/routes/agents.ts
import { Router, Request, Response } from "express";
import { requireUser } from "../middleware/auth";
import { prisma } from "../prisma";

const router = Router();

// Retrieve list of public/featured agents
router.get("/", requireUser, async (req: Request, res: Response) => {
  const { category, featured } = req.query;

  try {
    const whereClause: any = {
      OR: [
        { isPublic: true },
        { authorId: req.user?.id },
      ],
    };

    if (category) {
      whereClause.category = category as string;
    }

    if (featured === "true") {
      whereClause.isFeatured = true;
    }

    const agents = await prisma.agent.findMany({
      where: whereClause,
      orderBy: { usageCount: "desc" },
    });

    // Map to AIAgent frontend type structure
    const mappedAgents = agents.map((a) => ({
      id: a.id,
      name: a.name,
      description: a.description,
      systemPrompt: a.systemPrompt,
      modelId: a.modelId,
      category: a.category,
      iconEmoji: a.iconUrl || "🤖", // fallback to robot emoji
      rating: a.rating,
      usageCount: a.usageCount,
      isPublic: a.isPublic,
      createdBy: a.authorId,
      tags: [a.category],
      createdAt: a.createdAt.toISOString(),
    }));

    return res.json(mappedAgents);
  } catch (error: any) {
    console.error("Fetch agents failed:", error);
    return res.status(500).json({ detail: error.message });
  }
});

// Create custom workspace agent
router.post("/", requireUser, async (req: Request, res: Response) => {
  if (!req.user) return res.status(401).json({ detail: "Unauthorized" });

  const { name, description, systemPrompt, modelId, category, iconEmoji } = req.body;

  if (!name || !description || !systemPrompt || !modelId) {
    return res.status(400).json({ detail: "name, description, systemPrompt, and modelId are required" });
  }

  try {
    const agent = await prisma.agent.create({
      data: {
        name,
        description,
        systemPrompt,
        modelId,
        authorId: req.user.id,
        category: category || "general",
        iconUrl: iconEmoji || "🤖",
        isPublic: false,
        isFeatured: false,
      },
    });

    return res.status(201).json({
      id: agent.id,
      name: agent.name,
      description: agent.description,
      systemPrompt: agent.systemPrompt,
      modelId: agent.modelId,
      category: agent.category,
      iconEmoji: agent.iconUrl || "🤖",
      rating: agent.rating,
      usageCount: agent.usageCount,
      isPublic: agent.isPublic,
      createdBy: agent.authorId,
      tags: [agent.category],
      createdAt: agent.createdAt.toISOString(),
    });
  } catch (error: any) {
    console.error("Create agent failed:", error);
    return res.status(500).json({ detail: error.message });
  }
});

// Modify agent configurations
router.patch("/:id", requireUser, async (req: Request, res: Response) => {
  if (!req.user) return res.status(401).json({ detail: "Unauthorized" });

  const { id } = req.params;
  const { name, description, systemPrompt, modelId, category, iconEmoji } = req.body;

  try {
    const agent = await prisma.agent.findUnique({ where: { id } });
    if (!agent) {
      return res.status(404).json({ detail: "Agent profile not found" });
    }

    if (agent.authorId !== req.user.id && !req.user.isAdmin) {
      return res.status(403).json({ detail: "Permission denied" });
    }

    const updated = await prisma.agent.update({
      where: { id },
      data: {
        name: name !== undefined ? name : agent.name,
        description: description !== undefined ? description : agent.description,
        systemPrompt: systemPrompt !== undefined ? systemPrompt : agent.systemPrompt,
        modelId: modelId !== undefined ? modelId : agent.modelId,
        category: category !== undefined ? category : agent.category,
        iconUrl: iconEmoji !== undefined ? iconEmoji : agent.iconUrl,
        updatedAt: new Date(),
      },
    });

    return res.json({
      id: updated.id,
      name: updated.name,
      description: updated.description,
      systemPrompt: updated.systemPrompt,
      modelId: updated.modelId,
      category: updated.category,
      iconEmoji: updated.iconUrl || "🤖",
      rating: updated.rating,
      usageCount: updated.usageCount,
      isPublic: updated.isPublic,
      createdBy: updated.authorId,
      tags: [updated.category],
      createdAt: updated.createdAt.toISOString(),
    });
  } catch (error: any) {
    console.error("Update agent failed:", error);
    return res.status(500).json({ detail: error.message });
  }
});

// Remove workspace custom agent
router.delete("/:id", requireUser, async (req: Request, res: Response) => {
  if (!req.user) return res.status(401).json({ detail: "Unauthorized" });

  const { id } = req.params;

  try {
    const agent = await prisma.agent.findUnique({ where: { id } });
    if (!agent) {
      return res.status(404).json({ detail: "Agent profile not found" });
    }

    if (agent.authorId !== req.user.id && !req.user.isAdmin) {
      return res.status(403).json({ detail: "Permission denied" });
    }

    await prisma.agent.delete({ where: { id } });
    return res.status(204).send();
  } catch (error: any) {
    console.error("Delete agent failed:", error);
    return res.status(500).json({ detail: error.message });
  }
});

export default router;
